beacon-chain/db/slasherkv/pruning.go

```diff
@@ -0,0 +1,206 @@
+)
+
+// PruneProposals prunes all proposal data older than currentEpoch - historyLength in
+// specified epoch increments. BoltDB cannot handle long-running transactions, so we instead
+// use a cursor-based mechanism to prune at pruningEpochIncrements by opening individual bolt
+// transactions for each pruning iteration.
+func (s *Store) PruneProposals(
+	ctx context.Context, currentEpoch, pruningEpochIncrements, historyLength types.Epoch,
+) error {
+	if currentEpoch < historyLength {
+		log.Debugf("Current epoch %d < history length %d, nothing to prune", currentEpoch, historyLength)
+		return nil
+	}
+	// We can prune everything less than the current epoch - history length.
+	endEpoch := currentEpoch - historyLength
+	endPruneSlot, err := helpers.StartSlot(endEpoch)
+	if err != nil {
+		return err
+	}
+	encodedEndPruneSlot := fssz.MarshalUint64([]byte{}, uint64(endPruneSlot))
+
+	// We retrieve the lowest stored slot in the proposals bucket.
+	var lowestSlot types.Slot
+	if err = s.db.View(func(tx *bolt.Tx) error {
+		proposalBkt := tx.Bucket(proposalRecordsBucket)
+		c := proposalBkt.Cursor()
+		k, _ := c.First()
+		lowestSlot = slotFromProposalKey(k)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// If the lowest slot is greater than or equal to the end pruning slot,
+	// there is nothing to prune, so we return early.
+	if lowestSlot >= endPruneSlot {
+		log.Debugf("Lowest slot %d is >= pruning slot %d, nothing to prune", lowestSlot, endPruneSlot)
+		return nil
+	}
+
+	// We prune in increments of `pruningEpochIncrements` at a time to prevent
+	// a long-running bolt transaction which overwhelms CPU and memory.
+	// While we still have epochs to prune based on a cursor, we continue the pruning process.
+	epochAtCursor := helpers.SlotToEpoch(lowestSlot)
+	for endEpoch-epochAtCursor > 0 {
+		// Each pruning iteration involves a unique bolt transaction. Given pruning can be
+		// a very expensive process which puts pressure on the database, we perform
+		// the process in a batch-based method using a cursor to proceed to the next batch.
+		log.Debugf("Pruned %d/%d epochs worth of proposals", epochAtCursor, endEpoch-1)
+		if err = s.db.Update(func(tx *bolt.Tx) error {
+			proposalBkt := tx.Bucket(proposalRecordsBucket)
+			c := proposalBkt.Cursor()
+
+			var lastPrunedEpoch, epochsPruned types.Epoch
+			// We begin a pruning iteration at starting from the first item in the bucket.
+			for k, _ := c.First(); k != nil; k, _ = c.Next() {
+				// We check the slot from the current key in the database.
+				// If we have hit a slot that is greater than the end slot of the pruning process,
+				// we then completely exit the process as we are done.
+				if !uint64PrefixLessThan(k, encodedEndPruneSlot) {
+					epochAtCursor = endEpoch
+					return nil
+				}
+				// If we have pruned N epochs in this pruning iteration,
+				// we exit from the bolt transaction.
+				if epochsPruned >= pruningEpochIncrements {
+					return nil
+				}
```

Comment:
Should we move this check, right at the end of the loop, after
```golang
slasherProposalsPrunedTotal.Inc()
```
that will feel more natural (plus we can always be sure that `epochAtCursor` is updated (so our outer loop will receive updated value when we reach enough pruned epochs).

My understanding is:
- You move cursor, get some key
- Do checks (`uint64PrefixLessThan` etc)
- Delete key
- Update stats (including `epochsPruned++`
- At the end of the loop decide whether bold transaction should exit (and return if it should, but all the values, especially `epochAtCursor` are in the latest, updated state)

Reply:
Yep, that makes perfect sense and tests still passed